doorman/jwt_generic.go

```diff
@@ -0,0 +1,189 @@
+	}
+	return config, nil
+}
+
+func (v *jwtGenericValidator) jwks() (*JWKS, error) {
+	cacheKey := "jwks:" + v.Issuer
+	data, err := v.cache.Get(cacheKey)
+
+	// Cache is empty or expired: fetch again.
+	if err != nil {
+		config, err := v.config()
+		if err != nil {
+			return nil, err
+		}
+		uri := config.JWKSUri
+		log.Debugf("Fetch public keys from %s", uri)
+		data, err = downloadJSON(uri)
+		if err != nil {
+			return nil, err
+		}
+		v.cache.Set(cacheKey, data)
+	}
+
+	// XXX: since cache stores bytes, we parse it again at every usage :( ?
+	var jwks = &JWKS{}
+	err = json.Unmarshal(data, jwks)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(jwks.Keys) < 1 {
+		return nil, fmt.Errorf("No JWKS found")
+	}
+	return jwks, nil
+}
+
+func (v *jwtGenericValidator) ValidateRequest(r *http.Request) (*Claims, error) {
+	// 1. Extract JWT from request headers
+	token, err := fromHeader(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. Read JWT headers
+	if len(token.Headers) < 1 {
+		return nil, fmt.Errorf("No headers in the token")
+	}
+	header := token.Headers[0]
+	if header.Algorithm != string(v.SignatureAlgorithm) {
+		return nil, fmt.Errorf("Invalid algorithm")
+	}
+
+	// 3. Get public key with specified ID
+	keys, err := v.jwks()
+	if err != nil {
+		return nil, err
+	}
+	var key *jose.JSONWebKey
+	for _, k := range keys.Keys {
+		if k.KeyID == header.KeyID {
+			key = &k
+			break
+		}
+	}
+	if key == nil {
+		return nil, fmt.Errorf("No JWT key with id %q", header.KeyID)
+	}
+
+	// 4. Parse and verify signature.
+	jwtClaims := jwt.Claims{}
+	err = token.Claims(key, &jwtClaims)
+	if err != nil {
+		return nil, err
+	}
+
+	// 5. Validate issuer, claims and expiration.
+	// Will check audience only when request comes in, leave empty for now.
+	audience := []string{}
+	expected := jwt.Expected{Issuer: v.Issuer, Audience: audience}
+	expected = expected.WithTime(time.Now())
+	err = jwtClaims.Validate(expected)
+	if err != nil {
+		return nil, err
+	}
+
+	// 6. Extract relevant claims for Doorman.
+	claims, err := v.ClaimExtractor.Extract(token, key)
+	if err != nil {
+		return nil, err
+	}
+	return claims, nil
+}
+
+// fromHeader reads the authorization header value and parses it as JSON Web Token.
+func fromHeader(r *http.Request) (*jwt.JSONWebToken, error) {
+	if authorizationHeader := r.Header.Get("Authorization"); len(authorizationHeader) > 7 && strings.EqualFold(authorizationHeader[0:7], "BEARER ") {
+		raw := []byte(authorizationHeader[7:])
+		return jwt.ParseSigned(string(raw))
+	}
+	return nil, fmt.Errorf("Token not found")
+}
+
+func downloadJSON(uri string) ([]byte, error) {
+	response, err := http.Get(uri)
+	if err != nil {
+		return nil, err
+	}
+	if contentHeader := response.Header.Get("Content-Type"); !strings.HasPrefix(contentHeader, "application/json") {
+		return nil, fmt.Errorf("%s has not a JSON content-type", uri)
+	}
+	defer response.Body.Close()
+	data, err := ioutil.ReadAll(response.Body)
```

Comment:
check that `response.Body != nil`, otherwise it might panic.

Reply:
```
        // The http Client and Transport guarantee that Body is always
        // non-nil, even on responses without a body or responses with
        // a zero-length body. 
```

Reply:
My mistake. I was thinking about http.Request where `nil` means the Request has no body.